apache/hive
# HIVE-26569: Support renewal and recreation of LLAP_TOKENs

### What changes were proposed in this pull request?
This change implements a periodic LLAP_TOKEN token renewal and cancellation/recreation.
Both renewal and cancellation/recreation happens in the LLAP daemons, and it works on-demand. TezAM can call for a delegation token and it can authenticate itself by kerberos login and an existing token. Kerberos login support is needed when a token is cancelled, because in that situation that's is the only way to authenticate (in the absence of tokens). That's expected in case of long-running applications.


### Why are the changes needed?
To support LLAP_TOKEN usage beyond the lifetime given in creation time.

### Does this PR introduce _any_ user-facing change?
Yes, users can set below configs to control delegation token lifetime:
```
hive.llap.daemon.delegation.token.lifetime
hive.llap.daemon.delegation.token.renew.interval
```


### How was this patch tested?
Tested on a secure cluster with the following scenarios:
1. TezAM + Daemon, both healthy, running periodic check TezAM->Daemon
2. Close to token renewal time, Daemon renews token via SecretManager -> everything works as expected
3. Close to token maxLifetime, Daemon cancels token
a) if TezAM calls the periodic token refresh before other communication to Daemon, token is re-fetched, everything works as expected
b) if TezAM doesn't call periodic token refresh before other communication to Daemon (e.g. new DAG starts), there is a task attempt failure, but the tokens are refreshed before running the next attempt -> next attempt works fine with the new tokens as expected
4. Shutting down daemon:
a) token cancellation happens according to the logs as expected (-> cleaner and happier zookeeper)
b) TezAM tries to fetch a new token, tested with a single Daemon: token fetch is successful once Daemon is back and alive
